src/gurt/telemetry.c

```diff
@@ -682,6 +707,86 @@ d_tm_build_path(char **path, char *item, va_list args)
 	return rc;
 }
 
+/**
+ * Compute standard deviation
+ *
+ * \param[in]	sum_of_squares	Precomputed sum of squares
+ * \param[in]	sample_size	Number of elements in the data set
+ * \param[in]	mean		Mean of all elements
+ *
+ * \return			computed standard deviation
+ */
+double
+d_tm_compute_standard_dev(double sum_of_squares, uint64_t sample_size,
+			  double mean)
+{
+	if (sample_size == 0)
+		return 0;
+
+	return sqrtl((sum_of_squares - (sample_size * mean * mean)) /
+		     (sample_size - 1));
+}
+
+/**
+ * Compute gauge statistics: sample size, min, max, sum and sum of squares.
+ * Standard deviation calculation is deferred until the metric is read.
+ *
+ * \param[in]	node		Pointer to a gauge node
+ */
+void
+d_tm_compute_gauge_stats(struct d_tm_node_t *node)
+{
+	struct d_tm_stats_t	*dtm_stats;
+	uint64_t		value = 0;
+
+	dtm_stats = node->dtn_metric->dtm_stats;
+
+	if (dtm_stats == NULL)
+		return;
+
+	value = node->dtn_metric->dtm_data.value;
+	dtm_stats->sample_size++;
+	dtm_stats->dtm_sum.sum_int += value;
+	dtm_stats->sum_of_squares += value * value;
+
+	if (value > dtm_stats->dtm_max.max_int)
+		dtm_stats->dtm_max.max_int = value;
+
+	if (value < dtm_stats->dtm_min.min_int)
+		dtm_stats->dtm_min.min_int = value;
+}
+
+/**
+ * Compute duration statistics: sample size, min, max, sum and sum of squares.
+ * Standard deviation calculation is deferred until the metric is read.
+ *
+ * \param[in]	node		Pointer to a duration node
+ */
+void
+d_tm_compute_duration_stats(struct d_tm_node_t *node)
+{
+	struct d_tm_stats_t	*dtm_stats;
+	double			value = 0;
+
+	dtm_stats = node->dtn_metric->dtm_stats;
+
+	if (dtm_stats == NULL)
+		return;
+
+	value = node->dtn_metric->dtm_data.tms[0].tv_sec +
+		(node->dtn_metric->dtm_data.tms[0].tv_nsec / 1E9);
```

Comment:
I think that it would be more efficient to store it in nsec and avoid using double and a division here. Can be done in a subsequent patch.

Reply:
I am not following what you had in mind for storing the value in tv_nsec.  I did try my hand at a fixed point implementation to eliminate the divide within d_tm_compute_duration_stats().  What do you think of this instead?

```

void
d_tm_compute_duration_stats(struct d_tm_node_t *node)
{
	struct d_tm_stats_t	*dtm_stats;
	uint64_t		value = 0;

	dtm_stats = node->dtn_metric->dtm_stats;

	if (dtm_stats == NULL)
		return;

	value = (node->dtn_metric->dtm_data.tms[0].tv_sec * 1E9) +
		 node->dtn_metric->dtm_data.tms[0].tv_nsec;

	dtm_stats->sample_size++;
	dtm_stats->dtm_sum.sum_int += value;
	dtm_stats->sum_of_squares += (long double)value * (long double)value;

	if (value > dtm_stats->dtm_max.max_int)
		dtm_stats->dtm_max.max_int = value;

	if (value < dtm_stats->dtm_min.min_int)
		dtm_stats->dtm_min.min_int = value;
}

```

Then when the data is read, it is converted back to float inside d_tm_get_duration() shown in this snippet:

```

		if (stats != NULL) {
			stats->dtm_min.min_float = (double)dtm_stats->dtm_min.min_int / 1E9;
			stats->dtm_max.max_float = (double)dtm_stats->dtm_max.max_int / 1E9;
			stats->dtm_sum.sum_float = (double)dtm_stats->dtm_sum.sum_int / 1E9;
			if (dtm_stats->sample_size > 0)
				stats->mean = stats->dtm_sum.sum_float /
					      dtm_stats->sample_size;
			stats->sum_of_squares = (double)((dtm_stats->sum_of_squares / 1E18));
			stats->std_dev = d_tm_compute_standard_dev(
						      stats->sum_of_squares,
						      dtm_stats->sample_size,
						      stats->mean);

			stats->sample_size = dtm_stats->sample_size;
		}

```

Reply:
Looks good. Integer multiplication calculation is faster than floating point division.